### ISSUE TITLE
Configuration option for Java client OAuth credential cache type

### ISSUE THREAD

**Issue by wollefitz** [camunda/camunda#13124]
**Is your feature request related to a problem? Please describe.**

When using the Java client, OAuth tokens will be cached in a filesystem location as implemented in #3009. While this makes sense in most cases, it's not always ideal.

For example, when running a Dockerized application that uses `zeebe-client-java`, OAuth tokens will be written to `$HOME/.camunda/credentials` by default. If the Docker container is ran as non-root user (e.g. by setting Kubernetes' `securityContext.runAsUser` or using the image on Openshift), users will stumble into the following error:

```
java.nio.file.AccessDeniedException: /.camunda
	at java.base/sun.nio.fs.UnixException.translateToIOException(Unknown Source)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(Unknown Source)
	at java.base/sun.nio.fs.UnixFileSystemProvider.createDirectory(Unknown Source)
	at java.base/java.nio.file.Files.createDirectory(Unknown Source)
	at java.base/java.nio.file.Files.createAndCheckIsDirectory(Unknown Source)
	at java.base/java.nio.file.Files.createDirectories(Unknown Source)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsCache.ensureCacheFileExists(OAuthCredentialsCache.java:107)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsCache.writeCache(OAuthCredentialsCache.java:69)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.refreshCredentials(OAuthCredentialsProvider.java:133)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.loadCredentials(OAuthCredentialsProvider.java:121)
	at io.camunda.zeebe.client.impl.oauth.OAuthCredentialsProvider.applyCredentials(OAuthCredentialsProvider.java:79)
	at io.camunda.zeebe.client.impl.ZeebeCallCredentials.lambda$applyRequestMetadata$0(ZeebeCallCredentials.java:48)
	... 3 common frames omitted
```

**Describe the solution you'd like**

I can choose between different types of credential caches for `zeebe-client-java`, e.g. `file` or `in-memory`.

**Describe alternatives you've considered**

The token cache location can be proactively created and assigned the necessary permissions in an application's Dockerfile (see for example https://github.com/camunda/web-modeler/blob/fa476ef21005468c65fe5c01208a1f0828ae6d0b/restapi/Dockerfile#L32-L34). While this works, it takes away the user's possibility to customize the token cache path via the environment variable `ZEEBE_CLIENT_CONFIG_PATH`.

**Additional context**

This issue emerged from a customer support request (see https://jira.camunda.com/browse/SUPPORT-16086?focusedId=285736&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-285736 and https://github.com/camunda/web-modeler/issues/4780).

**Comment by korthout** [comment]
ZPA triage:
- workaround is available (customer is unblocked in support case) - see described alternative
- we wonder whether access could be provided if the folder uses the XDG Base Directory Specification (perhaps OpenShift respects this spec)
- @aleksander-dytko Can you verify whether this is a valid feature request?
- Size is a bit unclear, we'll start with estimation on `medium`

**Comment by aleksander-dytko** [comment]
I believe the request makes sense, but it's not high priority - setting prio to `Later`

**Comment by rob2universe** [comment]
This issue creates significant effort with prospect/customers that have relevant security policies. The images should out-of-the box work without fs access and with nonRootUser. The helm charts should mount persistent volumes as needed and those should be used instead.

**Comment by korthout** [comment]
>The images should out-of-the box work without file system access and with nonRootUser or elevated privileges. The helm charts should mount persistent volumes as needed and those should be used instead (while there is no in-memory cache)

💭 This is not about the helm charts nor about images. This is about the `zeebe-java-client`, a Java library.

>The required workarounds are not documented

👍 I agree that we should document the workaround here as well

**Comment by rob2universe** [comment]
I understand that this ticket is about the java client only. However, the need for filesystem access has an impact on the images and therefore on the bigger picture and ultimate perception by the consumers. While it may seem a low priority when looking at the java client only, solving the bigger picture should have a higher priority. Which solution (changing images or helm, e.g. to include add. persistent volume, vs in-mem cache for the client lib) is preferable is a decision which should be made with consideration of the bigger picture.

**Comment by korthout** [comment]
For completeness, I think that the workaround is simply:
- to mount an empty dir as zeebe client's config path

```yml
apiVersion: apps/v1
kind: Deployment
spec:
  template:
    spec:
      containers:
        env:
          - name: ZEEBE_CLIENT_CONFIG_PATH
            value: /usr/local/restapi/credentials/credentials.txt
        volumeMounts:
        - mountPath: /usr/local/restapi/credentials
          name: zeebe-client-config-path
      volumes:
      - emptyDir: {}
        name: zeebe-client-config-path
      securityContext:
        runAsUser: 999
```

**Comment by megglos** [comment]
In https://forum.camunda.io/t/camunda-client-in-read-only-filesystems/67179 a user hit the issue in a containerized setup with read only fs.

I generally question the value of the file system cache for the credentials.

The only use case I see is to retain credentials across application restarts. It was initially done to safe costs on obtaining new tokens with https://github.com/camunda/camunda/issues/3009 so it seemed very Camunda SaaS specific.

In contrast we have seen it causing issues over time, corrupted caches with multiple local client instances requiring deletion etc.

So I'm rather in favor of making in memory the default and keep an opt-in file cache feature, for those who really want this.

**Comment by megglos** [comment]
Reopening as we need to revise docs
> Docs update in camunda/camunda-docs (follow-up).
> Announce change in 8.10 docs

**Comment by megglos** [comment]
docs PR is here https://github.com/camunda/camunda/issues/13124 will close eagerly as the link to the issue from docs PR doesn't close this issue automatically
